config.go

```diff
@@ -83,20 +84,25 @@ type Config struct {
 
 	LoggerFactory logging.LoggerFactory
 
-	// ConnectTimeout is the timeout threshold for new connection handshakes
-	// to complete (default is 30 seconds)
-	ConnectTimeout *time.Duration
+	// function to make a context used in Dial(), Client(), Server(), and Accept().
+	// If nil, DefaultConnectContextMaker is used.
+	ConnectContextMaker func() (context.Context, func())
 
 	// MTU is the length at which handshake messages will be fragmented to
 	// fit within the maximum transmission unit (default is 1200 bytes)
 	MTU int
 }
 
-const defaultConnectTimeout = 30 * time.Second
+// DefaultConnectContextMaker is a default ConnectContextMaker used in Dial(), Client(), Server(), and Accept().
+var DefaultConnectContextMaker = func() (context.Context, func()) {
```

Comment:
Mind making this a function? I don't believe modification is anything we will ever need!

Reply:
In stdlib, default things are often defined as global variable, e.g. `net.DefaultResolver`: https://golang.org/pkg/net/#pkg-variables

One of my intention of this was providing an example of ConnectContextMaker implementation on godoc.
I'll make it a function, unexpose, and explicitly add an example as a comment.

Reply:
Sorry I am wrong! Wasn't aware this was a Go pattern. I just always avoid globals and the possibility of people modifying things. 

 If you want to revert and go back to the way you had it I am happy with that :)

Reply:
updated

Reply:
But yea, it may allow something bad. If one of the library does
```go
func init() {
    net.DefaultResolver = &net.Resolver{
        Dial: EvilDial,
    }
}
```
It's weird.

Reply:
So I agree this is a pattern in stdlib, but I think this is a place where we should deviate from stdlib, and I'm happy to see the `defaultConnectContextMaker` instead.

I understand why it's historically shown up in stdlib, net/http does something similar with `DefaultHTTPClient` so you can call a naked `http.$Method` without needing to create your own `http.Client` first, but it's proven to be quite the footgun.